src/annotator/anchoring/pdf.js

```diff
@@ -446,57 +446,71 @@ function prioritizePages(position) {
  * @return {Promise<Range>}
  */
-export function anchor(root, selectors) {
-  const position = /** @type {TextPositionSelector|undefined} */ (
-    selectors.find(s => s.type === 'TextPositionSelector')
-  );
+export async function anchor(root, selectors) {
   const quote = /** @type {TextQuoteSelector|undefined} */ (
     selectors.find(s => s.type === 'TextQuoteSelector')
   );
 
-  /** @type {Promise<Range>} */
-  let result = Promise.reject('unable to anchor');
+  // The quote selector is required in order to check that text position
+  // selector results are still valid.
+  if (!quote) {
+    throw new Error('No quote selector found');
+  }
+
+  const position = /** @type {TextPositionSelector|undefined} */ (
+    selectors.find(s => s.type === 'TextPositionSelector')
+  );
 
   if (position) {
-    result = result.catch(() => {
-      return findPage(position.start).then(({ index, offset, textContent }) => {
-        const start = position.start - offset;
-        const end = position.end - offset;
-        const length = end - start;
-
-        const matchedText = textContent.substr(start, length);
-        if (quote && quote.exact !== matchedText) {
-          throw new Error('quote mismatch');
-        }
+    // If we have a position selector, try using that first as it is the fastest
+    // anchoring method.
+    try {
+      const { index, offset, textContent } = await findPage(position.start);
+      const start = position.start - offset;
+      const end = position.end - offset;
+      const length = end - start;
+
+      const matchedText = textContent.substr(start, length);
+      if (quote.exact !== matchedText) {
+        throw new Error('quote mismatch');
+      }
 
-        return anchorByPosition(index, start, end);
-      });
-    });
-  }
+      const range = await anchorByPosition(index, start, end);
+      return range;
+    } catch {
+      // Fall back to quote selector
+    }
 
-  if (quote) {
-    result = result.catch(() => {
+    // If anchoring with the position failed, check for a cached quote-based
+    // match using the quote + position as a cache key.
+    try {
       if (
-        position &&
         quotePositionCache[quote.exact] &&
         quotePositionCache[quote.exact][position.start]
       ) {
         const { pageIndex, anchor } =
           quotePositionCache[quote.exact][position.start];
-        return anchorByPosition(pageIndex, anchor.start, anchor.end);
+        const range = await anchorByPosition(
+          pageIndex,
+          anchor.start,
+          anchor.end
+        );
+        return range;
       }
-
-      return prioritizePages(position?.start ?? 0).then(pageIndices => {
-        return findInPages(pageIndices, quote, position);
-      });
-    });
+    } catch {
+      // Fall back to uncached quote selector match
+    }
   }
 
-  return result;
+  const pageIndices = await prioritizePages(position?.start ?? 0);
+  return findInPages(pageIndices, quote, position);
```

Comment:
Correct me if I am wrong, Robert: if the anchoring by position fails and starts searching the quote using `findInPages`, it has the potential to slow down the application? Would `findInPages` will fetch text in the neighbouring pages or in the whole document to find a match? (Sorry, I haven't look at the code).

Reply:
> Correct me if I am wrong, Robert: if the anchoring by position fails and starts searching the quote using findInPages, it has the potential to slow down the application?

You're right. Searching by quote is slower than searching by position. The most significant overheads are:

1. Getting PDF.js to extract text from additional pages in the document. We mitigate this by caching the text for each page for future calls to `anchor`.
2. Doing an approximate string search for the quote. We mitigate this by first trying to find an exact match, using `string.indexOf(...)`, which is very fast. If we do need to fall back to a fuzzy/approximate search, we rely on the approximate searching implementation being highly optimized.

Of these, (1) is the most slowest part. Fortunately PDF.js does this work in a web worker, so the browser is not unresponsive while it is happening. We may want to show some kind of activity/progress while it is happening in very long documents however.